crates/zir/src/intern.rs

```diff
@@ -0,0 +1,110 @@
+//! Interned types for efficient memory usage and fast comparisons
+
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::ptr;
+
+use parking_lot::RwLock;
+use rustc_hash::FxHashSet;
+
+/// An interned reference to a value. Equality is pointer-based.
+#[repr(transparent)]
+pub struct Interned<'a, T: ?Sized>(&'a T);
+
+impl<'a, T: ?Sized> Interned<'a, T> {
+    #[inline]
+    pub const fn new_unchecked(value: &'a T) -> Self {
+        Interned(value)
+    }
+
+    #[inline]
+    pub fn as_ptr(self) -> *const T {
+        self.0 as *const T
+    }
+}
+
+impl<'a, T: ?Sized> Clone for Interned<'a, T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<'a, T: ?Sized> Copy for Interned<'a, T> {}
+
+impl<'a, T: ?Sized> Deref for Interned<'a, T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &T {
+        self.0
+    }
+}
+
+impl<'a, T: ?Sized> PartialEq for Interned<'a, T> {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        ptr::eq(self.0, other.0)
+    }
+}
+
+impl<'a, T: ?Sized> Eq for Interned<'a, T> {}
+
+impl<'a, T: ?Sized> Hash for Interned<'a, T> {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        ptr::hash(self.0, state)
+    }
+}
+
+impl<'a, T: fmt::Debug + ?Sized> fmt::Debug for Interned<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl<'a, T: fmt::Display + ?Sized> fmt::Display for Interned<'a, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+/// A set for interning values. Thread-safe.
+pub struct InternSet<'a, T: Hash + Eq> {
+    set: RwLock<FxHashSet<&'a T>>,
+}
```

Comment:
Instead of RwLock use this implementation:
https://doc.rust-lang.org/stable/nightly-rustc/src/rustc_data_structures/sharded.rs.html#24